In [1]:
import pandas as pd
import openpyxl

from scipy import stats
from sklearn.model_selection import train_test_split
from sklearn.impute import SimpleImputer
from pydataset import data
from sklearn.metrics import recall_score, accuracy_score, precision_score

In [2]:
#creating a rubric for the metrics

data = {
    '': ['Actual Positive', 'Actual Negative', ' ', 'Accuracy', 'Precision', 'Recall'] ,
    'Predicted Positive': ['True Positive (TP)', 'False Positive (FP)', ' ', '(TP + TN) /', 'TP /', 'TP / '],
    'Predicted Negative': ['False Negative (FN)', 'True Negative (TN)', ' ', '(TP + TN + FP + FN)', '(TP + FP)', '(TP + FN) '], 
}

rubric = pd.DataFrame(data)
rubric.set_index('', inplace=True)
rubric

Predicted Positive   Predicted Negative
                                                         
Actual Positive   True Positive (TP)  False Negative (FN)
Actual Negative  False Positive (FP)   True Negative (TN)
                                                         
Accuracy                 (TP + TN) /  (TP + TN + FP + FN)
Precision                       TP /            (TP + FP)
Recall                         TP /            (TP + FN)

# 2) 
Given the following confusion matrix, evaluate (by hand) the model's performance.

In [3]:
#creating table

data = ({'pred dog': ['46', '13'], 'pred cat': ['7', '34']})
index_labels = ['actual dog', 'actual cat']
dc_df = pd.DataFrame(data, index_labels)
dc_df

pred dog pred cat
actual dog       46        7
actual cat       13       34

a) In the context of this problem, what is a false positive?

False positive is predicting that there are 13 dogs when in actuality they were cats.

b) In the context of this problem, what is a false negative?

False negative is predicting that there are 7 cats but in actuality they were dogs.

c) How would you describe this model?

Evaluating our baseline!

Accuracy = (46 + 34) / (46 + 34 + 13 + 7)  
ans=> 4 / 5

Presision = 46 / (46 + 13)  
ans=> 46 / 59

Recall = 46 / (46 + 7)   
ans=> 46/53

# 3)
You are working as a datascientist working for Codeup Cody Creator (C3 for short), a rubber-duck manufacturing plant.

Unfortunately, some of the rubber ducks that are produced will have defects. Your team has built several models that try to predict those defects, and the data from their predictions can be found here.

Use the predictions dataset and pandas to help answer the following questions:

In [4]:
df_google = pd.read_excel('google_train.xlsx')

In [5]:
df = pd.read_csv('c3.csv')
df.head()

actual     model1  model2     model3
0  No Defect  No Defect  Defect  No Defect
1  No Defect  No Defect  Defect     Defect
2  No Defect  No Defect  Defect  No Defect
3  No Defect     Defect  Defect     Defect
4  No Defect  No Defect  Defect  No Defect

a) An internal team wants to investigate the cause of the manufacturing defects. They tell you that they want to identify as many of the ducks that have a defect as possible. 

In [6]:
#grouping all No Defect and Defects from all columns

act_nd = (df.actual == 'No Defect').sum()
act_d = (df.actual == 'Defect').sum()

mod1_nd = (df.model1 == 'No Defect').sum()
mod1_d = (df.model1 == 'Defect').sum()

mod2_nd = (df.model2 == 'No Defect').sum()
mod2_d = (df.model2 == 'Defect').sum()

mod3_nd = (df.model3 == 'No Defect').sum()
mod3_d = (df.model3 == 'Defect').sum()

# model 1

In [7]:
#creating a crosstab of the actual and model1 columns

pd.crosstab(df.actual, df.model1)

model1     Defect  No Defect
actual                      
Defect          8          8
No Defect       2        182

Evaluating model 1!

Accuracy = (8 + 182) / (8 + 182 + 2 + 8)   
ans=> 0.95

Presision =  8 / (8 + 2)   
ans=> 0.8

Recall = 8 / (8 + 8)     
ans=> 0.5

# model 2

In [8]:
#creating a crosstab of the actual and model2 columns

pd.crosstab(df.actual, df.model2)

model2     Defect  No Defect
actual                      
Defect          9          7
No Defect      81        103

Evaluating model 2!

Accuracy = (9 + 103) / (9 + 103 + 81 + 7)  
ans=> 0.56

Precision = 9 / (9 + 81)  
ans=> 0.1

Recall = 9 / (9 + 7)  
ans=> 0.56

# model 3

In [9]:
#creating a crosstab of the actual and model3 columns

pd.crosstab(df.actual, df.model3)

model3     Defect  No Defect
actual                      
Defect         13          3
No Defect      86         98

Evaluating model 3!

Accuracy = (13 + 98) / (13 + 98 + 86 + 3)  
ans=> 0.56

Precision = 13 / (13 + 86)  
ans=> 0.13

Recall = 13 / (13 + 3)  
ans=> 0.81

Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?

i) Recall will be the best metric to use in this case because to correctly identify defective ducks we will need to focus on minimizing as much false negatives(defective ducks predicted as non-defective) as possible.

ii) The best model to use in the case will be model 3 because it has the highest recall prediction percentage amongst all three models.

b) Recently several stories in the local news have come out highlighting customers who received a rubber duck with a defect, 

and portraying C3 in a bad light. The PR team has decided to launch a program that gives customers with a defective duck a vacation to Hawaii. 

They need you to predict which ducks will have defects, but tell you the really don't want to accidentally give out a vacation package when the duck really doesn't have a defect.

Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?

i) Precision will be the best metric to use in this case because it minimizes all the false positive outcomes and   
we don't want C3 to give out free vacations to customers that got non-defective ducks

ii) The best model to use in this case will be model 1 becasue it has the highest precision prediction percentage amongst all three models.

# 4)
You are working as a data scientist for Gives You Paws ™, a subscription based service that shows you cute pictures of dogs or cats (or both for an additional fee).

At Gives You Paws, anyone can upload pictures of their cats or dogs. The photos are then put through a two step process.  
First an automated algorithm tags pictures as either a cat or a dog (Phase I). Next, the photos that have been initially identified are put through   another round of review, possibly with some human oversight, before being presented to the users (Phase II).

Given this dataset, use pandas to create a baseline model (i.e. a model that just predicts the most common class) and answer the following questions:


In [10]:
gyp_df = pd.read_csv('gives_you_paws.csv')
gyp_df.head()

actual model1 model2 model3 model4
0    cat    cat    dog    cat    dog
1    dog    dog    cat    cat    dog
2    dog    cat    cat    cat    dog
3    dog    dog    dog    cat    dog
4    cat    cat    cat    dog    dog

a) In terms of accuracy, how do the various models compare to the baseline model? Are any of the models better than the baseline?

# model 1

In [11]:
#creating a crosstab of the actual and model1 columns

pd.crosstab(gyp_df.actual, gyp_df.model1)

model1   cat   dog
actual            
cat     1423   323
dog      640  2614

Evaluating model 1!

Accuracy =  (1423 + 2614) / (1423 + 2614 + 323 + 640)  
ans=> 0.80

Presision(cats) =  1423 / (1423 + 640)  
ans=> 0.69     
Precision(dogs) = 2614 / (2614 + 323)  
ans=>0.89

Recall(cats) =  1423 / (1423 + 323)  
ans=> 0.82  
Recall(dogs) = 2614 / (2614 + 640)  
ans=> 0.80

# model 2

In [12]:
#creating a crosstab of the actual and model2 columns

pd.crosstab(gyp_df.actual, gyp_df.model2)

model2   cat   dog
actual            
cat     1555   191
dog     1657  1597

Evaluating model 2!

Accuracy =  (1555 + 1597) / (1555 + 191 + 1657 + 1597)  
ans=> 0.52 

Precision(cats) = 1555 / (1555 + 1657)  
ans=>0.48    
Precision(dogs) = 1597 / (1597 + 191)  
ans=> 0.89

Recall(cats) = 1555 / (1555 + 191)   
ans=> 0.89  
Recall(dogs) = 1597 / (1597 + 1657)  
ans=> 0.49

# model 3

In [13]:
#creating a crosstab of the actual and model3 columns

pd.crosstab(gyp_df.actual, gyp_df.model3)

model3   cat   dog
actual            
cat      893   853
dog     1599  1655

Evaluating model 3!

Accuracy = (893 + 1655) / (893 + 853 + 1599 + 1655)  
ans=> 0.52

Precision(cats) = 893 / (893 + 1599)   
ans=> 0.36  
Precision(dogs) = 1655 / (1655 + 853)  
ans=> 0.66

Recall = 893 / (893 + 853)    
ans=> 0.51  
Recall = 1655 / (1655 + 1599)  
ans=> 0.51

# model 4

In [14]:
#creating a crosstab of the actual and model4 columns

pd.crosstab(gyp_df.actual, gyp_df.model4)

model4  cat   dog
actual           
cat     603  1143
dog     144  3110

In [15]:
3110 / (3110 + 144)

0.9557467732022127

Evaluating model 3!

Accuracy = (603 + 3110) / (603 + 1143 + 144 + 3110)  
ans=> 0.75 

Precision(cats) = 603 / (603 + 144)   
ans=> 0.81  
Precision(dogs) = 3110 / (3110 + 1143)  
ans=> 0.73  

Recall(cats) = 603 / (603 + 1143)   
ans=> 0.35  
Recall(dogs) = 3110 / (3110 + 144)  
ans=> 0.96

# baseline model

In [16]:
#creating a basline column

mode = gyp_df.actual.mode()

gyp_df['baseline_model'] = 'dog'
gyp_df.head()

actual model1 model2 model3 model4 baseline_model
0    cat    cat    dog    cat    dog            dog
1    dog    dog    cat    cat    dog            dog
2    dog    cat    cat    cat    dog            dog
3    dog    dog    dog    cat    dog            dog
4    cat    cat    cat    dog    dog            dog

In [17]:
#creating a crosstab of the actual and baseline columns

pd.crosstab(gyp_df.actual, gyp_df.baseline_model)

baseline_model   dog
actual              
cat             1746
dog             3254

#evaluating the baseline model

Accuracy = 3254 / (1746 + 3254)  
ans=> 0.65 

Precision = 3254 / 3254   
ans=> 1.0

Recall = 3254 / (3254 + 0)   
ans=> 1.0

In terms of accuracy model 1 has the highest percentage amongst all the models.   
Also model1 and model4 both have higher accuracy prediction percentages than the baseline model.

b) Suppose you are working on a team that solely deals with dog pictures. Which of these models would you recommend?

For the recall metric Model 4 will be the best model because of it's 96% prediction percentage.

For the precision metric Model 1 and 2 will be the best models because they both have an 89% prediction percentage.

c) Suppose you are working on a team that solely deals with cat pictures. Which of these models would you recommend?

For the recall metric Model 2 will be the best model because of it's 89% prediction percentage.

For the precision metric model 4 will be the best model becasue of it's 81% prediction percentage.